File: src/main/java/com/mindscapehq/android/raygun4android/network/RaygunNetworkLogger.java
package main.java.com.mindscapehq.android.raygun4android.network;

import java.net.URL;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import main.java.com.mindscapehq.android.raygun4android.RaygunClient;
import main.java.com.mindscapehq.android.raygun4android.RaygunLogger;
import main.java.com.mindscapehq.android.raygun4android.RaygunPulseEventType;
import main.java.com.mindscapehq.android.raygun4android.RaygunSettings;
import main.java.com.mindscapehq.android.raygun4android.network.http.RaygunUrlStreamHandlerFactory;

public class RaygunNetworkLogger {

  private static final long CONNECTION_TIMEOUT = 60000L; // 1 min
  private static volatile HashMap<String, RaygunNetworkRequestInfo> connections = new HashMap<String, RaygunNetworkRequestInfo>();
  private static boolean loggingEnabled = true;
  private static boolean loggingInitialized = false;

  public static void init() {
    if (loggingEnabled && !loggingInitialized) {
      try {
        RaygunUrlStreamHandlerFactory factory = new RaygunUrlStreamHandlerFactory();
        RaygunLogger.d("MD | Setting URL Stream Handler Factory!");
        URL.setURLStreamHandlerFactory(factory);
        loggingInitialized = true;
      }
      catch (SecurityException e) {
        loggingInitialized = false;
      }
    }
  }

  public static void setEnabled(boolean enabled) {
    loggingEnabled = enabled;
  }

  public static synchronized void startNetworkCall(String url, long startTime) {
    RaygunLogger.d("MD | startNetworkCall: "+url);
    if (!shouldIgnoreUrl(url)) {

        String id = sanitiseUrl(url);
        connections.put(id, new RaygunNetworkRequestInfo(url, startTime));

        removeOldEntries();
    }
  }

  public static synchronized void endNetworkCall(String url, long endTime, int statusCode) {
    if (url != null) {

      String id = sanitiseUrl(url);
      if ((connections.containsKey(id))) {

        RaygunNetworkRequestInfo request = connections.get(id);
        if (request != null) {
          connections.remove(url);
          sendNetworkTimingEvent(request.url, request.startTime, endTime, statusCode, null);
        }
      }
    }
  }

  public static synchronized void cancelNetworkCall(String url, long endTime, String exception) {
    if (url != null) {
      String id = sanitiseUrl(url);
      if ((connections != null) && (connections.containsKey(id))) {
        RaygunNetworkRequestInfo request = connections.get(id);
        if (request != null) {
          connections.remove(id);
          sendNetworkTimingEvent(request.url, request.startTime, endTime, 0, exception);
        }
      }
    }
  }

  public static synchronized void sendNetworkTimingEvent(String url, long startTime, long endTime, int statusCode, String exception) {
    if (!shouldIgnoreUrl(url)) {
      RaygunClient.SendPulseTimingEvent(RaygunPulseEventType.NetworkCall, url, endTime - startTime);
    }
  }

  private static synchronized void removeOldEntries() {
    Iterator<Map.Entry<String, RaygunNetworkRequestInfo>> it = connections.entrySet().iterator();
    while (it.hasNext()) {
      Map.Entry<String, RaygunNetworkRequestInfo> pairs = it.next();
      long startTime = pairs.getValue().startTime;
      if (System.currentTimeMillis() - startTime > CONNECTION_TIMEOUT) {
        it.remove();
      }
    }
  }

  private static String sanitiseUrl(String url) {
    if (url != null) {
      url = url.toLowerCase();
      url = url.replaceAll("https://", "");
      url = url.replaceAll("http://", "");
      url = url.replaceAll("www.", "");
    }
    return url;
  }

  private static boolean shouldIgnoreUrl(String url) {
    if (url == null) {
      return true;
    }
    for (String ignoredUrl : RaygunSettings.getSettings().getIgnoredUrls()) {
      if (url.contains(ignoredUrl) || ignoredUrl.contains(url)) {
        RaygunLogger.d("MD | RaygunNetworkLogger - Ignoring: "+url);
        return true;
      }
    }
    return false;
  }
}

File: provider/src/main/java/com/raygun/raygun4android/network/RaygunNetworkLogger.kt
package com.raygun.raygun4android.network

import com.raygun.raygun4android.RaygunRUMEventType
import com.raygun.raygun4android.RaygunSettings
import com.raygun.raygun4android.network.http.RaygunUrlStreamHandlerFactory
import com.raygun.raygun4android.rum.RUM
import java.net.URL

object RaygunNetworkLogger {
    private const val CONNECTION_TIMEOUT = 60000L // 1 min
    private val connections = HashMap<String?, RaygunNetworkRequestInfo>()
    private var loggingEnabled = true
    private var loggingInitialized = false

    @JvmStatic
    fun init() {
        if (loggingEnabled && !loggingInitialized) {
            try {
                val factory = RaygunUrlStreamHandlerFactory()
                URL.setURLStreamHandlerFactory(factory)
                loggingInitialized = true
            } catch (e: SecurityException) {
                loggingInitialized = false
            }
        }
    }

    @JvmStatic
    fun setEnabled(enabled: Boolean) {
        loggingEnabled = enabled
    }

    @Synchronized
    fun startNetworkCall(
        url: String?,
        startTime: Long,
    ) {
        if (!shouldIgnoreURL(url) && loggingEnabled) {
            removeOldEntries()
            val id = sanitiseURL(url)
            connections[id] = RaygunNetworkRequestInfo(url, startTime)
        }
    }

    @Synchronized
    fun endNetworkCall(
        url: String?,
        requestMethod: String,
        endTime: Long,
        statusCode: Int,
    ) {
        if (url != null) {
            val id = sanitiseURL(url)
            if ((connections.containsKey(id))) {
                val request = connections[id]
                if (request != null) {
                    sendNetworkTimingEvent(
                        request.url,
                        requestMethod,
                        request.startTime,
                        endTime,
                        statusCode,
                        null,
                    )
                    connections.remove(id)
                }
            }
        }
    }

    /**
     * When a network request is cancelled we stop tracking it and do not send the information
     * through. Future updates may include sending the cancelled request timing through with
     * information showing it was cancelled.
     *
     * @param url URL to cancel
     * @param requestMethod URL to cancel
     * @param endTime URL to cancel
     * @param exception URL to cancel
     */
    @Synchronized
    fun cancelNetworkCall(
        url: String?,
        requestMethod: String?,
        endTime: Long,
        exception: String?,
    ) {
        if (url != null) {
            val id = sanitiseURL(url)
            connections.remove(id)
        }
    }

    @Synchronized
    private fun sendNetworkTimingEvent(
        url: String?,
        requestMethod: String,
        startTime: Long,
        endTime: Long,
        statusCode: Int,
        exception: String?,
    ) {
        if (!shouldIgnoreURL(url) && loggingEnabled) {
            val sanitizedUrl = sanitiseURL(url)
            RUM
                .getInstance()
                .sendRUMTimingEvent(
                    RaygunRUMEventType.NETWORK_CALL,
                    "$requestMethod $sanitizedUrl",
                    endTime - startTime,
                )
        }
    }

    @Synchronized
    private fun removeOldEntries() {
        val it: MutableIterator<Map.Entry<String?, RaygunNetworkRequestInfo>> =
            connections.entries.iterator()
        while (it.hasNext()) {
            val pairs = it.next()
            val startTime = pairs.value.startTime
            if (System.currentTimeMillis() - startTime > CONNECTION_TIMEOUT) {
                it.remove()
            }
        }
    }

    private fun sanitiseURL(url: String?): String? {
        var sanitizedUrl = url
        if (sanitizedUrl != null) {
            val queryIndex = sanitizedUrl.indexOf("?")
            if (queryIndex > 0) {
                sanitizedUrl = sanitizedUrl.substring(0, queryIndex)
            }
        }
        return sanitizedUrl
    }

    private fun shouldIgnoreURL(url: String?): Boolean {
        if (url == null) {
            return true
        }
        for (ignoredUrl in RaygunSettings.getIgnoredURLs()) {
            if (url.contains(ignoredUrl) || ignoredUrl.contains(url)) {
                return true
            }
        }
        return false
    }
}
